### PR TITLE
DBZ-302 Add support for `database.whitelist` mongodb connector config…

### PR DESCRIPTION
**Pull request by ekreiser**
DBZ-302 Add support for `database.whitelist` mongodb connector configuration option

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java [hunk 1]
```diff
@@ -250,7 +252,7 @@ protected boolean performInitialSync() {
         // We need to copy each collection, so put the collection IDs into a queue ...
         final List<CollectionId> collections = new ArrayList<>();
         primaryClient.collections().forEach(id -> {
-            if (collectionFilter.test(id)) collections.add(id);
+            if (databaseFilter.test(id.dbName()) && collectionFilter.test(id))collections.add(id);
```
Comment (rhauch) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java]: Minor nit: watch formatting and the space after the if condition.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java [hunk 2]
```diff
@@ -430,6 +432,7 @@ protected boolean handleOplogEvent(ServerAddress primaryAddress, Document event)
                     logger.error("Get current primary executeBlocking", e);
                 }
                 ServerAddress serverAddress = address.get();
+
```
Comment (rhauch) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java]: While you're making changes, it'd be great to avoid unrelated changes like this.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 3]
```diff
@@ -162,6 +162,18 @@
                                                    .withDependents(COLLECTION_LIST_NAME)
                                                    .withDescription("The databases for which changes are to be captured");
 
+        /**
+     * A comma-separated list of regular expressions that match the fully-qualified namespaces of collections to be monitored.
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: This comment seems like a Copy & Paste glitch, I don't think it should mention collections here.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 4]
```diff
@@ -162,6 +162,18 @@
...
+     * Fully-qualified namespaces for collections are of the form {@code '<databaseName>.<collectionName>'}.
+     * May not be used with {@link #COLLECTION_BLACKLIST}.
+     */
+    public static final Field DATABASE_WHITELIST = Field.create("database.whitelist")
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: It would be nice to have a `database.blacklist`, too, so to resemble the collection-level options.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java [hunk 5]
```diff
@@ -22,13 +22,21 @@
...
      * Create an instance of the filters.
      * 
      * @param config the configuration; may not be null
      */
     public Filters(Configuration config) {
+        String dbWhitelist = config.getString(MongoDbConnectorConfig.DATABASE_WHITELIST);
+        if (dbWhitelist != null && !dbWhitelist.trim().isEmpty()){
+            databaseFilter = Predicates.includes(dbWhitelist);
+        } else {
+            databaseFilter = Predicates.includes(".*");
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java]: Can we pass a simple predicate which always returns true instead of passing it through the regex engine?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java (hunk 1 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java]: space added

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java (hunk 2 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java]: extra line removed

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 3 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: Corrected the comment

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 4 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: Added database.blacklist support

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java (hunk 5 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java]: changed to use simple predicate when a pattern isn't set

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java [hunk 6]
```diff
@@ -22,13 +22,24 @@
...
 
     /**
      * Create an instance of the filters.
      * 
      * @param config the configuration; may not be null
      */
     public Filters(Configuration config) {
+        String dbWhitelist = config.getString(MongoDbConnectorConfig.DATABASE_WHITELIST);
+        String dbBlacklist = config.getString(MongoDbConnectorConfig.DATABASE_BLACKLIST);
+        if (dbWhitelist != null && !dbWhitelist.trim().isEmpty()){
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java]: I am sorry but it seems that space is missing between rightmost parenthesis and the left curl brace following

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 7]
```diff
@@ -252,4 +279,14 @@ private static int validateCollectionBlacklist(Configuration config, Field field
         }
         return 0;
     }
+    
+    private static int validateDatabaseBlacklist(Configuration config, Field field, ValidationOutput problems) {
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: It seems we have an interesting inconsistency here - MySQL connector supports adding both blacklist and whitelist but PostgreSQL requires setting of only one of those options. I'd vote going MySQL way and retrofit the functionality into Postgres connector too. Let Gunnar decide.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java (hunk 6 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Filters.java]: corrected

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 7 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: I see your point but was trying to be consistent within the mongo connector.  The collection.whitelist and collection.blacklist are currently coded to be mutually exclusive (not sure why).  If the current collection setting were not coded this way, I would be able to combine collection.whitelist and collection.blacklist patterns that accomplish what I need without ever needing the settings specifically at the database level (possible since the collection comparisons are performed while including the database name as a prefix).

Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: So it turns out, combining includes and exludes aren't supported in the MySQL connector atm. either. Apparently it was meant to be supported at some point (there's some code in place for merging them), but then this validation was added (which you added to MongoDB, too) which prevents giving both options. So I'd suggest we do this for now (i.e. your PR is good as is), and then, as a follow-up task, we allow merging includes + excludes across all the connectors. WDYT?

Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: There's https://issues.jboss.org/browse/DBZ-307 for the follow-up task.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java [hunk 8]
```diff
@@ -199,15 +224,17 @@
...
+                                                     DATABASE_BLACKLIST);
 
     protected static Field.Set EXPOSED_FIELDS = ALL_FIELDS;
 
     protected static ConfigDef configDef() {
         ConfigDef config = new ConfigDef();
         Field.group(config, "MongoDB", HOSTS, USER, PASSWORD, LOGICAL_NAME, CONNECT_BACKOFF_INITIAL_DELAY_MS,
                     CONNECT_BACKOFF_MAX_DELAY_MS, MAX_FAILED_CONNECTIONS, AUTO_DISCOVER_MEMBERS);
-        Field.group(config, "Events", DATABASE_LIST, COLLECTION_WHITELIST, COLLECTION_BLACKLIST);
+        Field.group(config, "Events", DATABASE_LIST, DATABASE_WHITELIST, DATABASE_BLACKLIST, COLLECTION_WHITELIST, COLLECTION_BLACKLIST);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: Hum, so that's a bit weird, there already was `DATABASE_LIST` before, but just its definition, I can't seem to find code that would read that property, also it's not part of the docs. Seems like an uncompleted attempt of adding an option for specifying the DBs. So I'd suggest to remove this with this PR as well, as it's superseded by the new options now.

**Comment by gunnarmorling** [comment]
Hey @ekreiser, so I think the PR s good, apart from my last comment about removing that left-over `DATABASE_LIST` option. Could you also send in a PR for updating [the docs](https://github.com/debezium/debezium.github.io/blob/develop/docs/connectors/mongodb.asciidoc) to cover the new options, too?

Thanks!

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java (hunk 7 above)
Comment (ekreiser) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java]: I think it is a good plan to separate this issue out and look at it holistically cross all connectors.

**Comment by ekreiser** [comment]
I added DBZ-309 to track the task of updating the documentation to include the new configuration settings included by this PR.

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks a lot, @ekreiser! I took the liberty to squash some of the commits and remove some trailing whitespace inserted accidentally. I'll send a PR for updating the docs, too.
